storage/shard.go

```diff
@@ -357,13 +361,19 @@ func (s *dbShard) forEachShardEntry(entryFn dbShardEntryWorkFn) error {
 	s.RLock()
 	elemsLen := s.list.Len()
 	batchSize := int(math.Ceil(shardIterateBatchPercent * float64(elemsLen)))
-	nextElem := s.list.Front()
 	s.RUnlock()
 
-	// TODO(xichen): pool or cache this.
-	currEntries := make([]*dbShardEntry, 0, batchSize)
-	for nextElem != nil {
+	var (
+		currEntries = make([]*dbShardEntry, 0, batchSize) // TODO(xichen): pool or cache this.
+		first       = true
+		nextElem    *list.Element
+	)
+	for nextElem != nil || first {
 		s.RLock()
+		if first {
+			nextElem = s.list.Front()
+			first = false
+		}
```

Comment:
If you wanted to get cute, this seems to work:


Checked using the playground:

Reply:
Missing the paste/links here